llvm/lib/SYCLLowerIR/SYCLLinkedModuleProcessor.cpp

```diff
@@ -0,0 +1,45 @@
+//===-- SYCLLinkedModuleProcessor.cpp - finalize a fully linked module ---===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLLinkedModuleProcessor.h"
+
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "sycl-linked-module-processor"
+using namespace llvm;
+
+namespace {
+class SYCLLinkedModuleProcessor : public ModulePass {
+public:
+  static char ID;
+  SYCLLinkedModuleProcessor(SpecConstantsPass::HandlingMode Mode)
+      : ModulePass(ID), Mode(Mode) {
+    initializeSYCLLinkedModuleProcessorPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    // TODO: determine if we need to run other passes
```

Comment:
If I understand correctly, that's an equivalent of what's being run by `sycl-post-link` after device code split is performed. If so, then we have the following other transformations applied at this stage:
- ESIMD handling, which includes some special module fixup for `invoke_simd`, as well as potential additional split by ESIMD followed up by optional linking that back
- Generation of a separate device image with default values of spec constants

If we also taking about what happens after `llvm-link` but before device code split, then it is also:
- Something about `invoke_simd`
- Sanitizer-related passes
- Joint matrix passes

Reply:
So when we do early splitting in `-c` we actually run sycl-post-link in full, including all those passes. So in that case, we only need to run passes here that need the fully linked module. If we decide to change the design such that we do only split in `-c` but no passes, then we would need every pass that sycl-post-link runs. In the current implementation ~2100/2000 E2E tests are passing, so it seems most passes don't need the full module, at least for the test cases we have.

Reply:
> In the current implementation ~2100/2000 E2E tests are passing, so it seems most passes don't need the full module, at least for the test cases we have.

I believe that most of E2E are single-file tests with no `SYCL_EXTERNAL` dependencies. Even SYCL-CTS won't help you here. I suppose that we need more or less real-life applications here to be sure and gather more data if we need it

Reply:
I suggest adding tests similar to `sycl/test-e2e/Basic/multisource.cpp`.